### PR TITLE
Convert some images to multi-stage builds

### PR DESCRIPTION
**Pull request by rneugeba**
This converts `tini`, `media-toybox`, `rngd`, and `test-virtsock` to use the new multi-stage build. It also moves `tini` and `media-toybox` to `./tools` which seems a more appropriate place.

`c-compile` has also been extended to allow it to be used to compile the above C code

### CODE REVIEW

**Comment by avsm** [comment]
This doesn't work yet on the Hub autobuilds does it?

**Comment by justincormack** [comment]
@avsm no not yet, but we are not using those

File: pkg/rngd/Dockerfile [hunk 1]
```diff
@@ -1,3 +1,47 @@
...
+  linux-headers \
+  make \
+  musl-dev \
+  patch \
+  && true
+
+COPY . /
+
+ENV pkgname=rng-tools pkgver=5
+RUN curl -O -sSL http://downloads.sourceforge.net/project/gkernel/$pkgname/$pkgver/$pkgname-$pkgver.tar.gz
```
Comment (justincormack) [inline pkg/rngd/Dockerfile]: sourceforge...

File: pkg/rngd/Dockerfile [hunk 2]
```diff
@@ -1,3 +1,47 @@
...
+  strip /usr/sbin/rngd
+RUN mkdir -p /tmp/dev /tmp/proc /tmp/sys
+RUN cat ${HASHFILES} | sha1sum | sed 's/ .*//' > hash
+
+
+FROM linuxkit/tini:a847727dd242b65975e6e5e007bcfe31fbd30fef@sha256:587b2af6eb9a0657c51fccd496a910b06fd71bb14e932a9bd24f21e0f63315a1
+COPY --from=rng-build usr/sbin/rngd usr/sbin/rngd
+COPY --from=rng-build hash hash
+# Can't create directories, so copy empty dirs from previous context
+COPY --from=rng-build tmp/dev dev
```
Comment (justincormack) [inline pkg/rngd/Dockerfile]: why cant you just `RUN mkdir ...`?

**Comment by justincormack** [comment]
My plan was to get rid of the whole hashing thing by having a base Alpine image with all the packages available in a filesystem package repo (not installed, as its still good to see which ones are needed to be installed for each build).

File: pkg/rngd/Dockerfile (hunk 2 above)
Comment (rneugeba) [inline pkg/rngd/Dockerfile]: no `mkdir` in the `tini` image

Comment (justincormack) [inline pkg/rngd/Dockerfile]: ah

**Comment by rneugeba** [comment]
I'm not sure how a base Alpine image will help. It get's rid of some of the hashes, but not all. We still have images which are created based on input files and we need to detect if/when the input changed...

Also, let's do things one step at a time. I suggest we move to multi-stage builds first.

What's your view on unified (ie containing all necessary packages) c-compile image? Will make the dependent images simpler

**Comment by justincormack** [comment]
Yes, it will get rid of the non source changes.

What doesnt c-compile include?

**Comment by rneugeba** [comment]
Let me find out when I update :)

**Comment by justincormack** [comment]
But anyway this looks fine as a stage 1

**Comment by rneugeba** [comment]
Pushed an update with an updated `c-compile`. Simplifies the builds a lot.

File: Makefile [hunk 3]
```diff
@@ -71,16 +71,26 @@ test: test-initrd.img test-bzImage test-cmdline
 	tar cf - $^ | ./scripts/qemu.sh 2>&1 | tee test.log
 	$(call check_test_log, test.log)
 
+test-ltp.img.tar.gz: $(MOBY) test/ltp/test-ltp.yml
+	$(MOBY) build test/ltp/test-ltp.yml
```
Comment (justincormack) [inline Makefile]: these ltp parts are unrelated and already upstream...

File: pkg/rngd/Dockerfile (hunk 1 above)
Comment (justincormack) [inline pkg/rngd/Dockerfile]: we check the sha256sum instead

File: pkg/rngd/fix-textrels-on-PIC-x86.patch [hunk 4]
```diff
@@ -1,17 +1,17 @@
 --- rng-tools/rdrand_asm.S
 +++ rng-tools/rdrand_asm.S
 @@ -49,6 +49,7 @@
- 	ret
+	ret
```
Comment (justincormack) [inline pkg/rngd/fix-textrels-on-PIC-x86.patch]: why did this change?

File: src/cmd/moby/run_gcp.go [hunk 5]
```diff
@@ -45,6 +45,7 @@ func runGcp(args []string) {
 	familyFlag := gcpCmd.String("family", "", "GCP Image Family. A group of images where the family name points to the most recent image. *Optional* when 'prefix' is a filename")
 	nameFlag := gcpCmd.String("img-name", "", "Overrides the Name used to identify the file in Google Storage, Image and Instance. Defaults to [name]")
 	diskSizeFlag := gcpCmd.Int("disk-size", 0, "Size of system disk in GB")
+	skipCleanup := gcpCmd.Bool("skip-cleanup", false, "Don't remove images or VMs")
```
Comment (justincormack) [inline src/cmd/moby/run_gcp.go]: this shouldnt be in here either?

File: Makefile (hunk 3 above)
Comment (riyazdf) [inline Makefile]: yep - should just need a rebase for this and the gcp flags

File: pkg/rngd/fix-textrels-on-PIC-x86.patch (hunk 4 above)
Comment (rneugeba) [inline pkg/rngd/fix-textrels-on-PIC-x86.patch]: my git changed the whitespace, should be benign

Comment (rneugeba) [inline pkg/rngd/fix-textrels-on-PIC-x86.patch]: Hmm, more of this file changed. Let me check

Comment (rneugeba) [inline pkg/rngd/fix-textrels-on-PIC-x86.patch]: ah, looks like GH's web interface doesn't grok patch files very well. The changes are just whitespace changes, so should be fine
